app/application/services/merkle_service.py

```
@@ -0,0 +1,206 @@
+        path: list[MerkleProofStep] = []
+        while current_depth > 0:
+            parent_depth = current_depth - 1
+            parent_pos = current_pos // 2
+            parent = await self._merkle_repo.get_node(
+                epoch_id, parent_depth, parent_pos
+            )
+            if parent is None:
+                raise LookupError(
+                    f"Merkle tree incomplete: node missing at depth={parent_depth}, "
+                    f"position={parent_pos} (epoch_id={epoch_id!r})"
+                )
```

Ruff (0.15.5)

[warning] Avoid specifying long messages outside the exception class

(TRY003)